### ISSUE TITLE
The metrics for the last epoch in TensorBoard are being overridden.

### ISSUE THREAD

**Issue by kjhfkjn** [ultralytics/ultralytics#18001]
### Search before asking

- [X] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/ultralytics/ultralytics/discussions) and found no similar questions.

### Question

```
def final_eval(self):
        """Performs final evaluation and validation for object detection YOLO model."""
        ckpt = {}
        for f in self.last, self.best:
            if f.exists():
                if f is self.last:
                    ckpt = strip_optimizer(f)
                elif f is self.best:
                    k = "train_results"  # update best.pt train_metrics from last.pt
                    strip_optimizer(f, updates={k: ckpt[k]} if k in ckpt else None)
                    LOGGER.info(f"\nValidating {f}...")
                    self.validator.args.plots = self.args.plots
                    self.metrics = self.validator(model=f)
                    self.metrics.pop("fitness", None)
                    self.run_callbacks("on_fit_epoch_end")
```
In final_eval, self.run_callbacks("on_fit_epoch_end") causes the metrics of the last epoch to be overridden, resulting in an upward spike in TensorBoard at the end of each training
![屏幕截图 2024-12-04 164703](https://github.com/user-attachments/assets/bdb89b72-8b4b-4205-a6ba-a14a2526b492)

**Comment by UltralyticsAssistant** [comment]
👋 Hello @kjhfkjn, thank you for bringing this to our attention! 🚀 We appreciate your detailed breakdown and code snippet—it’s super helpful for understanding the context of your issue.

For 🐛 Bug Reports like this, we kindly request you to include a [minimum reproducible example (MRE)](https://docs.ultralytics.com/help/minimum_reproducible_example/) to help us efficiently debug the problem. This should ideally include:
1. A small, shareable script reproducing the issue.
2. Details about your environment (Python version, Ultralytics version, PyTorch version, etc.).
3. Clear reproduction steps, and if applicable, the dataset or portions of it relevant to the issue.

It seems like this issue might involve training logs and metrics visualization. Could you confirm if you're using the latest version of the `ultralytics` package? If not, please upgrade using the command below and verify if the issue persists:

```bash
pip install -U ultralytics
```

You can also test your setup in one of our verified environments to ensure compatibility:
- Notebooks with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo11"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>

For general training ❓ questions, we recommend reviewing our [Model Training Tips](https://docs.ultralytics.com/guides/model-training-tips/) as they cover many troubleshooting scenarios.

Finally, for community support, feel free to join the [Ultralytics Discord](https://discord.com/invite/ultralytics) 🎧 for real-time discussions or our [Discourse forum](https://community.ultralytics.com) for in-depth solutions. You can even check out our [Subreddit](https://reddit.com/r/Ultralytics) for engaging community threads.

📊 Quick Check: 
- Current Status: <a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yaml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yaml/badge.svg" alt="Ultralytics CI"></a>. If the CI badge is green, that indicates all official tests are currently passing.

This is an automated response, but no worries—an Ultralytics engineer will review your issue and provide assistance shortly. Thank you for your patience 😊!

**Comment by Y-T-G** [comment]
So is it showing the metrics of best.pt for last epoch?

**Comment by glenn-jocher** [comment]
@kjhfkjn the observed behavior occurs because the `final_eval` function in `run_callbacks("on_fit_epoch_end")` uses metrics from `best.pt` for re-validation during the last epoch, which reflects in TensorBoard. This doesn't influence early stopping, as it relies solely on the metrics during the actual training epochs. If you'd like to avoid this discrepancy in TensorBoard, modifying the callback logic to exclude re-validation at the final step may help.

**Comment by glenn-jocher** [comment]
Thank you for identifying and sharing this observation. We'll take this feedback into account to improve clarity in TensorBoard visualizations. You can also track updates or suggest enhancements by contributing to the repository.
